src/Compilers/CSharp/Test/Syntax/Diagnostics/LocationsTests.cs

```diff
@@ -194,32 +205,91 @@ public void TestInvalidLineMapping()
         {
             string sampleProgram = @"using System;
 class X {
-    int q; 
+    int q;
 #line 0 ""firstdirective""
-    int r; 
+    int r;
 #line 20 ""seconddirective""
-    int s; 
-}";
+    int s;
+}".NormalizeLineEndings();
+
             SyntaxTree syntaxTree = SyntaxFactory.ParseSyntaxTree(sampleProgram, path: "filename.cs");
 
             AssertMappedSpanEqual(syntaxTree, "int q", "filename.cs", 2, 4, 2, 9, hasMappedPath: false);
             AssertMappedSpanEqual(syntaxTree, "int r", "filename.cs", 4, 4, 4, 9, hasMappedPath: false); // invalid #line args
             AssertMappedSpanEqual(syntaxTree, "int s", "seconddirective", 19, 4, 19, 9, hasMappedPath: true);
+
+            AssertEx.Equal(new[]
+            {
+                "[|using System;\r\nclass X {\r\n    int q;\r\n|] -> : (0,0)-(2,12)",
+                "[|    int r;\r\n|] -> : (4,0)-(4,12)",
+                "[|    int s;\r\n}|] -> seconddirective: (19,0)-(20,1)"
+            }, InspectLineMapping(syntaxTree));
         }
 
         [Fact]
         public void TestLineMappingNoDirectives()
         {
             string sampleProgram = @"using System;
 class X {
-int x; 
+int x;
 }";
             SyntaxTree syntaxTree = SyntaxFactory.ParseSyntaxTree(sampleProgram, path: "c:\\goo.cs");
 
             AssertMappedSpanEqual(syntaxTree, "ing Sy", "c:\\goo.cs", 0, 2, 0, 8, hasMappedPath: false);
             AssertMappedSpanEqual(syntaxTree, "class X", "c:\\goo.cs", 1, 0, 1, 7, hasMappedPath: false);
             AssertMappedSpanEqual(syntaxTree, $"System;{Environment.NewLine}class X", "c:\\goo.cs", 0, 6, 1, 7, hasMappedPath: false);
             AssertMappedSpanEqual(syntaxTree, "x;", "c:\\goo.cs", 2, 4, 2, 6, hasMappedPath: false);
+
+            Assert.Empty(InspectLineMapping(syntaxTree));
+        }
+
+        [Fact]
+        public void TestLineMappingFirstAndLastLineDirectives()
+        {
+            string sampleProgram = @"#line 20
+class X {}
+#line 30".NormalizeLineEndings();
+            var syntaxTree = SyntaxFactory.ParseSyntaxTree(sampleProgram, path: "c:\\goo.cs");
+
+            AssertEx.Equal(new[]
+            {
+                "[|class X {}\r\n|] -> : (19,0)-(19,12)",
+            }, InspectLineMapping(syntaxTree));
+        }
+
+        [Fact]
+        public void TestLineMappingLastLineDirectiveFollowedByEmptyLine()
+        {
+            string sampleProgram = @"#line 30
+".NormalizeLineEndings();
+
+            var syntaxTree = SyntaxFactory.ParseSyntaxTree(sampleProgram, path: "c:\\goo.cs");
+
+            AssertEx.Equal(new[]
+            {
+                "[||] -> : (29,0)-(29,0)",
+            }, InspectLineMapping(syntaxTree));
+        }
+
+        [Fact]
+        public void TestLineMappingConsecutiveDirectives()
+        {
+            string sampleProgram =
+@"#line hidden
+#line default
+class C {}
+#line 5
+#line 6
```

Comment:
It's not clear, from the `AssertEx.Equal()` below, which of these line directives was ignored. Consider making the line numbers differ by some value other than 1.